src/extname.rs

```diff
@@ -0,0 +1,51 @@
+use std::path::MAIN_SEPARATOR;
+use libc::c_char;
+use std::ffi::{CStr, CString};
+use std::str;
+
+#[no_mangle]
+pub extern "C" fn extname(string: *const c_char) -> *const c_char {
+    let c_str = unsafe {
+        assert!(!string.is_null());
+
+        CStr::from_ptr(string)
+    };
+    let r_str = str::from_utf8(c_str.to_bytes()).unwrap_or("");
+
+    let mut chars_iter = r_str.char_indices().rev();
+
+    let mut reversed_ext: Vec<u8> = Vec::new();
+    let mut dot_index = 0;
+    let mut separator_index = 0;
+    loop {
+        let char = chars_iter.next();
+        if (char.is_some()) {
```

Comment:
It should be. I can't imagine why this generates any different code, unless there's a bug somewhere. How was performance being tested?